pkg/kademlia/kademlia_planet_test.go

```diff
@@ -86,3 +94,138 @@ func TestPingTimeout(t *testing.T) {
 
 	})
 }
+
+func TestBootstrapBackoffReconnect(t *testing.T) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	// This sets up an unreliable proxy server which will receive conns from
+	// storage nodes and the satellite, but doesn't connect them with
+	// the bootstrap node (proxy.target) until the droppedConnInterval has passed.
+	// This should test that the Bootstrap function will retry a connection
+	// if it initially fails.
+	proxy, err := newBadProxy(zaptest.NewLogger(t), "127.0.0.1:0")
+	require.NoError(t, err)
+
+	planet, err := testplanet.NewCustom(zaptest.NewLogger(t), testplanet.Config{
+		SatelliteCount: 1, StorageNodeCount: 4, UplinkCount: 0,
+		Reconfigure: testplanet.Reconfigure{
+			Satellite: func(log *zap.Logger, index int, config *satellite.Config) {
+				config.Kademlia.BootstrapAddr = proxy.listener.Addr().String()
+			},
+			StorageNode: func(index int, config *storagenode.Config) {
+				config.Kademlia.BootstrapAddr = proxy.listener.Addr().String()
+				config.Kademlia.BootstrapBackoffBase = 200 * time.Millisecond
+				config.Kademlia.BootstrapBackoffMax = 3 * time.Second
+			},
+		},
+	})
+	require.NoError(t, err)
+	defer ctx.Check(planet.Shutdown)
+
+	// We set the bad proxy's "target" to the bootstrap node's addr
+	// (which was selected when the new custom planet was set up).
+	proxy.target = planet.Bootstrap.Addr()
+
+	droppedConnInterval := 500 * time.Millisecond
+	go func() {
+		// This starts the unreliable proxy server and sets it to
+		// drop connections for the first 500 milliseconds that it's up.
+		err := proxy.run(droppedConnInterval)
+		require.NoError(t, err)
+	}()
+	defer func() {
+		err := proxy.close()
+		require.NoError(t, err)
+	}()
+
+	planet.Start(ctx)
+}
+
+type badProxy struct {
+	closed   uint32 // using sync/atomic package to do an atomic set/load of closed (as bool flag)
+	log      *zap.Logger
+	listener net.Listener
+	target   string
+}
+
+func newBadProxy(log *zap.Logger, addr string) (*badProxy, error) {
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &badProxy{
+		log:      log,
+		listener: l,
+	}, nil
+}
+
+func (proxy *badProxy) close() error {
+	atomic.StoreUint32(&proxy.closed, 1)
```

Comment:
To address Egon's concerns, I think you'd need to a waitgroup / errGroup in the struct (instead of simple go routines) and wait for everything to close here.

Reply:
I think everything is properly getting shut down now because of the atomic "closed" check, I think the problem still is that I'm just not propagating and checking all possible errors if I'm understanding Egon right.

Reply:
welp ended up using errGroup after all 😅thanks Bill